nixos/modules/services/networking/seafile.nix

```diff
@@ -0,0 +1,276 @@
+
+    ccnetSettings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+
+        options = {
+          General = {
+            SERVICE_URL = mkOption {
+              type = types.str;
+              example = "https://www.example.com";
+              description = ''
+                Seahub public URL.
+              '';
+            };
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for ccnet, see
+        <link xlink:href="https://manual.seafile.com/config/ccnet-conf/"/>
+        for supported values.
+      '';
+    };
+
+    seafileSettings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+
+        options = {
+          fileserver = {
+            port = mkOption {
+              type = types.port;
+              default = 8082;
+              description = ''
+                The tcp port used by seafile fileserver.
+              '';
+            };
+            host = mkOption {
+              type = types.str;
+              default = "127.0.0.1";
+              example = "0.0.0.0";
+              description = ''
+                The binding address used by seafile fileserver.
+              '';
+            };
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for seafile-server, see
+        <link xlink:href="https://manual.seafile.com/config/seafile-conf/"/>
+        for supported values.
+      '';
+    };
+
+    workers = mkOption {
+      type = types.int;
+      default = 4;
+      example = 10;
+      description = ''
+        The number of gunicorn worker processes for handling requests.
+      '';
+    };
+
+    adminEmail = mkOption {
+      example = "john@example.com";
+      type = types.str;
+      description = ''
+        Seafile Seahub Admin Account Email.
+      '';
+    };
+
+    initialAdminPassword = mkOption {
+      example = "someStrongPass";
+      type = types.str;
+      description = ''
+        Seafile Seahub Admin Account initial password.
+        Should be change via Seahub web front-end.
+      '';
+    };
+
+    seafilePackage = mkOption {
+      type = types.package;
+      description = "Which package to use for the seafile server.";
+      default = pkgs.seafile-server;
+    };
+
+    seahubSecretKeyFile = mkOption {
+      type = types.path;
+      description = ''
+        A file containing the django secret key for safe session.
+      '';
+    };
+
+    seahubExtraConf = mkOption {
+      default = "";
+      type = types.lines;
+      description = ''
+        Extra config to append to `seahub_settings.py` file.
+        Refer to <link xlink:href="https://manual.seafile.com/config/seahub_settings_py/" />
+        for all available options.
+      '';
+    };
+  };
+
+  ###### Implementation
+
+  config = mkIf cfg.enable {
+
+    environment.etc."seafile/ccnet.conf".source = ccnetConf;
```

Comment:
The idea was to have a centralized configuration directory fully managed by nix (see [this comment](https://github.com/NixOS/nixpkgs/pull/119719#discussion_r615796825) above).
Is there another way to cleanly (ie: no linking on startup/change) build such a directory?

Reply:
My approach has been to write the generated configs via`pkgs.writeText`, and copy them over (caution: permissions) to a common seafile state directory at each preExec, see https://git.orlives.de/schmittlauch/nixos-seafile-overlay/src/branch/mainline/mod-seafile-server.nix#L302

Reply:
I also thought about whether using the global environment of `/etc` just feels wrong & against the priniples of NixOS, or whether it also has some real disadvantages I can think of.
When putting all config files into `/etc`, it is only possible to run a single instance of the seafile service module at the same machine. While I'm not sure whether this is a commonly used scenario, achieving proper decoupling of multiple service instances is easily achievable here by putting all state and config into a configurable directory. No matter how relevant in practice, that's a good design principle anyways.
What @aanderse probably meant in his comment was the possibility of directly specifying the full path to the config file as an arument to seafile's start scripts, like done for multiple other services.

Reply:
We can achieve the same decoupling strategy with a config pattern like `/etc/seafile/{instance}/...`. Copying files is error prone and add unnecessary complexity.
While supporting multiple instances is a noble goal, I suggest we leave this out for another PR. For now let's try to have a running seafile :smile:

Reply:
Another option is to store the config files in `/var/lib/seafile` so that secrets can be substituted into them with `envsubst`. But I think using etc is fine too, there a reason the `environment.etc` option exists.

Reply:
> Another option is to store the config files in /var/lib/seafile so that secrets can be substituted into them with envsubst.

Oh thanks for the reminder. I indeed intend to extend this module with mysql database support, which indeed requires specifying secrets in config files. Is it possible to maintain the secret confidentiality by setting `environment.etc.<name>.mode` and `environment.etc.<name>.user` accordingly? IMHO secrets need to be substituted at service start time, because I know of no way to read the secrets from file at activation time.

If that is not possible in any sensible way, I need to ask you for the move to keeping config in the state dir. At last my PR on mysql support would need to include such a change otherwise.

Reply:
@greizgh I'd like to hear whether you intend to change this within this PR.
When introducing MySQL/MariaDB support, I need to adopt a solution that allows substituting secrets without storing them world-readable in nix store, and this is the approach I'd take anyways.

You can of course prefer to leave it as it is know and let me change the config location approach, or we can already do the correct thing from the start.

This is my last nit-pick before approving this PR-

Reply:
I prefer to use the `environment.etc` mechanism and avoid messing with copy/chown on init.
My understanding is that most services now accept a `dbSecretFile` to prevent storing credentials in nix store, I'm not sure how this relates to the use of `environment.etc`?

The strategy I went with for loading secrets is to fetch them from environment variables.

In `seahubExtraConf` for instance I have:
```python
import os
OAUTH_CLIENT_ID = os.getenv("OAUTH_CLIENT_ID")
OAUTH_CLIENT_SECRET = os.getenv("OAUTH_CLIENT_SECRET")
```
And use `systemd.services.seahub.serviceConfig.EnvironmentFile` to inject said env vars.
We could use the same pattern for DB credentials.

Reply:
While I am still of a different opinion, I am okay with going forward.
For future reference, seafile-server also accepts DB credentials from ENV vars: https://download.seafile.com/d/320e8adf90fa43ad8fee/files/?p=/docker/docker-compose.yml